app/jobs/export_pool_xp_to_traction_job.rb

```diff
@@ -0,0 +1,125 @@
+# frozen_string_literal: true
+
+# Combine the samples from a Pool XP tube into a compound sample, generate a bioscan-pool-xp-tube-to-traction message
+# and submit it to RabbitMQ so that it can forwarded to Traction by the message processor.
+ExportPoolXpToTractionJob =
+  Struct.new(:barcode) do
+    include CompoundSampleHelper
+
+    def perform
+      message_data = get_message_data(barcode)
+
+      subject_obj = configatron.amqp.schemas.subjects[:export_pool_xp_to_traction]
+      subject = subject_obj[:subject]
+      version = subject_obj[:version]
+
+      schema = get_message_schema(subject, version)
+      encoded_message = avro_encode_message(message_data, schema)
+      send_message(encoded_message, subject, version)
+    rescue StandardError => e
+      Rails.logger.error("Error exporting Pool XP tube to Traction: <#{e.message}>")
+      raise
+    end
+
+    def fetch(uri_str, limit = 10)
+      raise IOError, 'Too many HTTP redirects' if limit.zero?
+
+      response = Net::HTTP.get_response(URI.parse(uri_str))
+
+      case response
+        when Net::HTTPSuccess     then response
+        when Net::HTTPRedirection then fetch(response['location'], limit - 1)
+      else
+        response.error!
+      end
+    end
+
+    def get_message_data(barcode)
+      tube = Tube.with_barcode(barcode).first
+      project = tube.projects.first
+      study = tube.studies.first
+      sample = find_or_create_compound_sample(study, tube.samples)
+
+      {
+        messageUuid: UUIDTools::UUID.timestamp_create.to_s,
+        messageCreateDateUtc: Time.now.utc,
+        tubeBarcode: tube.human_barcode,
+        library: {
+          volume: 100.0,
+          concentration: 0.0,
+          boxBarcode: "Unspecified"
+        },
+        request: {
+          costCode: project.project_cost_code,
+          libraryType: "Pacbio_Amplicon",
+          studyUuid: study.uuid
+        },
+        sample: {
+          sampleName: sample.name,
+          sampleUuid: sample.uuid,
+          speciesName: "Unidentified"
+        }
+      }
+    end
+
+    def get_message_schema(subject, version)
+      response = fetch("#{configatron.amqp.schemas.registry_url}#{subject}/versions/#{version}")
+      resp_json = JSON.parse(response.body)
+      resp_json['schema']
+    end
+
+    def avro_encode_message(message, schema)
```

Comment:
I think these functions will be re-used; not sure they'll be re-used in sequencescape but surely will in other components (e.g. traction-service). Do you think these functions are worthy of being extracted into a library or something?

Reply:
Not yet, but I agree.  My usual rule of thumb is if you end up with three copies of the same code, it's time to refactor.  Any eariler, and you're introducing technical debt via abstraction for a theoretical future benefit.

See https://stackify.com/premature-optimization-evil/ and https://x.com/KentBeck/status/167620773337505793